core/modules/taxonomy/taxonomy.module

```diff
@@ -1852,6 +1853,53 @@ function taxonomy_field_widget_form(&$form, &$form_state, $field, $instance, $la
   return $element;
 }
 
+/**
+ * Implements hook_field_widget_form_alter().
+ */
+function taxonomy_field_widget_form_alter(&$element, &$form_state, $context) {
+  $field = $context['field'];
+  $widget_type = $context['instance']['widget']['type'];
+
+  if ($field['type'] == 'taxonomy_term_reference' && in_array($widget_type, array('options_buttons', 'options_select'))) {
+    if (empty($field['settings']['option_list_callback']) || $field['settings']['option_list_callback'] == 'taxonomy_allowed_values') {
```

Comment:
The correct key is `'options_list_callback'`, not `'option_list_callback'`.

HOWEVER, I'm thinking that all this `if / else` business is not needed. In the end, what you are doing here is simply adding the indentations to existing options stored   in `$element['#options']`. This means that we are not really worried about how those options are built (all access checks and translations happened earlier, when that options array was built). 

So, I think all is needed here is to simply do `$trees = $field['settings']['allowed_values'];` and remove all the `if/else` stuff below. I haven't tested this yet, but let me know if there is a reason why you need to do all this.

Reply:
The reason for the if/else was to save the extra db call to find the vocabularies for terms. If `allowed_values` is set and `taxonomy_allowed_values` is the callback, then we can get the vocabularies from `allowed_values`. But if there's a different callback, we can't count on anything in `allowed_values` (at least, that's what I understood you to say in some earlier comments), and so we have to get the vocabularies for all of the terms in order to get the term depths.

If we don't care about that extra db call, we can drop the if/else and just always get the vocabularies via the db call.

Reply:
Or conversely, if we can count on `allowed_values` even when there's a different callback, we can use it always. (That's what I was doing in an earlier version of this PR.)

Reply:
I'm so sorry @bugfolder. I think what I said earlier doesn't quite make sense. `allowed_values` will always contain the selected vocabularies, since that's determined in the field settings form. I got confused earlier, thinking that `allowed_values` actually contained the options - thus my confusion.

 > if we can count on allowed_values even when there's a different callback, we can use it always

Yes. No need to do a db call.

🙏🏽